sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala

```diff
@@ -133,7 +140,12 @@ class HDFSMetadataLog[T <: AnyRef : ClassTag](sparkSession: SparkSession, path:
           s"Multiple streaming queries are concurrently using $path", e)
       case e: Throwable =>
         output.cancel()
-        throw e
+        if (retryCount > metaDataNumRetries) {
+          throw new IOException(
+            s"Failed to write meta data log after retry $metaDataNumRetries count", e)
+        } else {
+          writeBatchToFileWithRetries(metadata, path, retryCount + 1)
```

Comment:
I'm pretty sure this recursive call is not needed.

Reply:
I add a UT but this just one case that close file error. Maybe there will have an other error.

Reply:
I understand what @gaborgsomogyi meant as use loop instead of recursive call, especially it's fairly easy to just use loop. This is even not a tail recursive so while I guess retryCount will be small number in practice it's not a good idea to do so.

Reply:
Oh, I got it too.